kani fails to compile a crate but cargo can build  it

I would like to use kani, but it fails to build one of my dependencies (`needletail`), even though `cargo build` runs fine.
Step to reproduce:
```bash
git clone https://github.com/onecodex/needletail
cd needletail
cargo build  # OK
```
However, kani fails to build:
```bash
cargo kani
# Kani Rust Verifier 0.57.0 (cargo plugin)
# [...]
# error: linking with `cc` failed: exit status: 1
#  = note: rust-lld: error: version script assignment of 'global' to symbol 'bz_internal_error' failed: symbol not defined
          collect2: error: ld returned 1 exit
# error: could not compile `needletail` (lib) due to 1 previous error
# error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

Note: when I remove `bzip2` from the dependencies of `needletail`, kani works fine, but I don't need to do it for `cargo build`.

Is there a difference in linking between `cargo build` and `cargo kani`?

Hi @carolynzech,
I am using amd64/Debian 12. I reproduced the bug with a Debian virtual machine (after installing packages: `git curl build-essential`).
I installed kani with:
```bash
cargo install --locked kani-verifier
cargo kani setup
```

Here is the output of `cargo kani --debug` on my (non virtual) machine. Do you need any other information?
```
Kani Rust Verifier 0.57.0 (cargo plugin)
2025-01-10T09:10:11.608742Z DEBUG kani_driver::call_cargo: packages_to_verify args package_selection=[] package_exclusion=[] workspace=false
2025-01-10T09:10:11.608756Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="needletail" kind=[CDyLib, RLib] crate_type=[CDyLib, RLib]
2025-01-10T09:10:11.608761Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="stdin_pipe" kind=[Example] crate_type=[Bin]
2025-01-10T09:10:11.608763Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="format_specimens" kind=[Test] crate_type=[Bin]
2025-01-10T09:10:11.608765Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="test_compressed" kind=[Test] crate_type=[Bin]
2025-01-10T09:10:11.608767Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="test_stdin" kind=[Test] crate_type=[Bin]
2025-01-10T09:10:11.608768Z DEBUG kani_driver::call_cargo: package_targets name="needletail" target="benchmark" kind=[Bench] crate_type=[Bin]
Skipped the following test targets: 'format_specimens', 'test_compressed', 'test_stdin'.
    -> Use '--tests' to verify harnesses inside a 'test' crate.
Skipped verification of the following unsupported targets: 'stdin_pipe', 'benchmark'.
[Kani] Running: `CARGO_ENCODED_RUSTFLAGS="-Coverflow-checks=on-Zunstable-options-Ztrim-diagnostic-paths=no-Zhuman_readable_cgu_names-Zalways-encode-mir--cfg=kani-Zcrate-attr=feature(register_tool)-Zcrate-attr=register_tool(kanitool)--sysroot/home/lrobidou/.kani/kani-0.57.0-L/home/lrobidou/.kani/kani-0.57.0/lib--externkani--externnoprelude:std=/home/lrobidou/.kani/kani-0.57.0/lib/libstd.rlib-Cpanic=abort-Csymbol-mangling-version=v0-Zpanic_abort_tests=yes-Zmir-enable-passes=-RemoveStorageMarkers--check-cfg=cfg(kani)--kani-compiler-Cllvm-args=--check-version=0.57.0 --log-level=debug --assertion-reach-checks" CARGO_TERM_PROGRESS_WHEN="never" RUSTC="/home/lrobidou/.kani/kani-0.57.0/bin/kani-compiler" /home/lrobidou/.kani/kani-0.57.0/toolchain/bin/cargo rustc --target x86_64-unknown-linux-gnu -Zhost-config -Ztarget-applies-to-host --config=host.rustflags=["--cfg=kani_host"] --target-dir /home/lrobidou/Documents/needletail/target/kani --message-format json-diagnostic-rendered-ansi -v -p path+file:///home/lrobidou/Documents/needletail#0.6.1 --lib -- -Cllvm-args=--reachability=harnesses`
       Fresh jobserver v0.1.32
       Fresh shlex v1.3.0
       Fresh cc v1.2.7
       Fresh pkg-config v0.3.31
       Fresh libc v0.2.169
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/liblibc-d1c61429ba3f4f30.rlib"
       Fresh zstd-sys v2.0.13+zstd.1.5.6
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd_sys-9f924567fce6184b.rlib"
       Fresh adler2 v2.0.0
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libadler2-cb22019072e7c439.rlib"
       Fresh cfg-if v1.0.0
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libcfg_if-181dde3da22bfac8.rlib"
       Fresh miniz_oxide v0.8.2
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libminiz_oxide-1a041dcadce55b49.rlib"
       Fresh bzip2-sys v0.1.11+1.0.8
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2_sys-ce23cb44da98297e.rlib"
       Fresh crc32fast v1.4.2
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libcrc32fast-ef8def920a2f839e.rlib"
       Fresh zstd-safe v7.2.1
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd_safe-b024a7887f9ab28a.rlib"
       Fresh liblzma-sys v0.3.11
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma_sys-c180c694e427ca62.rlib"
       Fresh memchr v2.7.4
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libmemchr-1df6418796d90927.rlib"
       Fresh bytecount v0.6.8
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbytecount-68dfcab22851b95b.rlib"
       Fresh flate2 v1.0.35
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libflate2-fb11999c1a5b09bd.rlib"
       Fresh buffer-redux v1.0.2
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbuffer_redux-7b4c530c02a6a1d5.rlib"
       Fresh zstd v0.13.2
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd-d4af044339cab2c4.rlib"
       Fresh bzip2 v0.4.4
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2-8dea9fe18d25e2b3.rlib"
       Fresh liblzma v0.3.5
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: None, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::codegen_cprover_gotoc::compiler_interface link, crate_type=Rlib, out_path="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma-53c7c9da0626f27c.rlib"
   Compiling needletail v0.6.1 (/home/lrobidou/Documents/needletail)
     Running `/home/lrobidou/.kani/kani-0.57.0/bin/kani-compiler --crate-name needletail --edition=2021 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=238 --crate-type cdylib --crate-type rlib --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="bzip2"' --cfg 'feature="compression"' --cfg 'feature="default"' --cfg 'feature="flate2"' --cfg 'feature="liblzma"' --cfg 'feature="xz2"' --cfg 'feature="zstd"' --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("bzip2", "compression", "default", "flate2", "liblzma", "pyo3", "python", "python_test", "xz2", "zstd"))' -C metadata=8124dbd093344b72 --out-dir /home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -C incremental=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/incremental -L dependency=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/lrobidou/Documents/needletail/target/kani/debug/deps --extern buffer_redux=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbuffer_redux-7b4c530c02a6a1d5.rlib --extern bytecount=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbytecount-68dfcab22851b95b.rlib --extern bzip2=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2-8dea9fe18d25e2b3.rlib --extern flate2=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libflate2-fb11999c1a5b09bd.rlib --extern liblzma=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma-53c7c9da0626f27c.rlib --extern memchr=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libmemchr-1df6418796d90927.rlib --extern zstd=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd-d4af044339cab2c4.rlib -Cllvm-args=--reachability=harnesses -C overflow-checks=on -Z unstable-options -Z trim-diagnostic-paths=no -Z human_readable_cgu_names -Z always-encode-mir --cfg=kani -Z 'crate-attr=feature(register_tool)' -Z 'crate-attr=register_tool(kanitool)' --sysroot /home/lrobidou/.kani/kani-0.57.0 -L /home/lrobidou/.kani/kani-0.57.0/lib --extern kani --extern 'noprelude:std=/home/lrobidou/.kani/kani-0.57.0/lib/libstd.rlib' -C panic=abort -C symbol-mangling-version=v0 -Z panic_abort_tests=yes -Z mir-enable-passes=-RemoveStorageMarkers '--check-cfg=cfg(kani)' --kani-compiler '-Cllvm-args=--check-version=0.57.0 --log-level=debug --assertion-reach-checks' -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/bzip2-sys-b726842da2f7dd95/out/lib -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/liblzma-sys-7733c1c74cb39892/out -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/zstd-sys-1e32c7efbbb876af/out`
DEBUG kani_compiler::kani_compiler config end, queries=QueryDb { args: Some(Arguments { check_assertion_reachability: true, check_coverage: false, emit_vtable_restrictions: false, output_pretty_json: false, ignore_global_asm: false, reachability_analysis: Harnesses, stubbing_enabled: false, unstable_features: [], build_std: false, log_level: Some(Directive { in_span: None, fields: [], target: None, level: LevelFilter::DEBUG }), json_output: false, color_output: false, check_version: Some("0.57.0"), ub_check: [], backend: CProver, print_llbc: false }), kani_functions: OnceCell(<uninit>) }
DEBUG kani_compiler::kani_middle::kani_functions find_kani_functions, kani=[Crate { id: 20, name: "kani", is_local: false }]
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(Assume), fn_def=FnDef(DefId { id: 56, name: "kani::assume" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(Assert), fn_def=FnDef(DefId { id: 57, name: "kani::assert" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(Cover), fn_def=FnDef(DefId { id: 58, name: "kani::cover" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(Any), fn_def=FnDef(DefId { id: 59, name: "kani::any" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(AnyModifies), fn_def=FnDef(DefId { id: 60, name: "kani::any_modifies" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(AnyRaw), fn_def=FnDef(DefId { id: 61, name: "kani::any_raw" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(Panic), fn_def=FnDef(DefId { id: 62, name: "kani::panic" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(SafetyCheck), fn_def=FnDef(DefId { id: 64, name: "kani::safety_check" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(UnsupportedCheck), fn_def=FnDef(DefId { id: 65, name: "kani::unsupported" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(UntrackedDeref), fn_def=FnDef(DefId { id: 66, name: "kani::internal::untracked_deref" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(InitContracts), fn_def=FnDef(DefId { id: 67, name: "kani::internal::init_contracts" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(WriteAny), fn_def=FnDef(DefId { id: 68, name: "kani::internal::write_any" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(WriteAnySlice), fn_def=FnDef(DefId { id: 69, name: "kani::internal::write_any_slice" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(WriteAnySlim), fn_def=FnDef(DefId { id: 70, name: "kani::internal::write_any_slim" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(WriteAnyStr), fn_def=FnDef(DefId { id: 71, name: "kani::internal::write_any_str" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(RunContract), fn_def=FnDef(DefId { id: 72, name: "kani::internal::run_contract_fn" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(RunLoopContract), fn_def=FnDef(DefId { id: 73, name: "kani::internal::run_loop_contract_fn" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(Check), fn_def=FnDef(DefId { id: 74, name: "kani::internal::check" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SizeOfVal), fn_def=FnDef(DefId { id: 153, name: "kani::rustc_intrinsics::size_of_val_raw" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(AlignOfVal), fn_def=FnDef(DefId { id: 154, name: "kani::rustc_intrinsics::align_of_val_raw" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(Offset), fn_def=FnDef(DefId { id: 155, name: "kani::rustc_intrinsics::offset" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SizeOfDynObject), fn_def=FnDef(DefId { id: 156, name: "kani::mem_models::size_of_dyn_object" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(AlignOfDynObject), fn_def=FnDef(DefId { id: 157, name: "kani::mem_models::align_of_dyn_object" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SizeOfSliceObject), fn_def=FnDef(DefId { id: 158, name: "kani::mem_models::size_of_slice_object" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(FloatToIntInRange), fn_def=FnDef(DefId { id: 159, name: "kani::float::float_to_int_in_range" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(CheckedSizeOf), fn_def=FnDef(DefId { id: 170, name: "kani::mem::checked_size_of_raw" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(CheckedAlignOf), fn_def=FnDef(DefId { id: 171, name: "kani::mem::checked_align_of_raw" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(IsAllocated), fn_def=FnDef(DefId { id: 172, name: "kani::mem::is_allocated" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(ValidValue), fn_def=FnDef(DefId { id: 173, name: "kani::mem::has_valid_value" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Intrinsic(IsInitialized), fn_def=FnDef(DefId { id: 174, name: "kani::mem::is_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(PointerObject), fn_def=FnDef(DefId { id: 175, name: "kani::mem::pointer_object" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Hook(PointerOffset), fn_def=FnDef(DefId { id: 176, name: "kani::mem::pointer_offset" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(InitializeMemoryInitializationState), fn_def=FnDef(DefId { id: 184, name: "kani::mem_init::initialize_memory_initialization_state" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(IsPtrInitialized), fn_def=FnDef(DefId { id: 185, name: "kani::mem_init::is_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SetPtrInitialized), fn_def=FnDef(DefId { id: 186, name: "kani::mem_init::set_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(IsSliceChunkPtrInitialized), fn_def=FnDef(DefId { id: 187, name: "kani::mem_init::is_slice_chunk_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SetSliceChunkPtrInitialized), fn_def=FnDef(DefId { id: 188, name: "kani::mem_init::set_slice_chunk_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(IsSlicePtrInitialized), fn_def=FnDef(DefId { id: 189, name: "kani::mem_init::is_slice_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SetSlicePtrInitialized), fn_def=FnDef(DefId { id: 190, name: "kani::mem_init::set_slice_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(IsStrPtrInitialized), fn_def=FnDef(DefId { id: 191, name: "kani::mem_init::is_str_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(SetStrPtrInitialized), fn_def=FnDef(DefId { id: 192, name: "kani::mem_init::set_str_ptr_initialized" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(CopyInitState), fn_def=FnDef(DefId { id: 193, name: "kani::mem_init::copy_init_state" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(CopyInitStateSingle), fn_def=FnDef(DefId { id: 194, name: "kani::mem_init::copy_init_state_single" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(StoreArgument), fn_def=FnDef(DefId { id: 196, name: "kani::mem_init::store_argument" })
DEBUG kani_compiler::kani_middle::kani_functions Found kani function, kani_function=Model(LoadArgument), fn_def=FnDef(DefId { id: 197, name: "kani::mem_init::load_argument" })
DEBUG kani_compiler::kani_middle::codegen_units CodegenUnits::new, units=[]
DEBUG kani_compiler::kani_middle::codegen_units store_metadata, filename="/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/needletail.kani-metadata.json"
====== Reachability Analysis Result =======
Total # items: 0
Total # statements: 0
Total # expressions: 0

Reachable Items:

Statements:

Expressions:

-------------------------------------------
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/lrobidou/.kani/kani-0.57.0/bin:/home/lrobidou/.kani/kani-0.57.0/pyroot/bin:/home/lrobidou/miniforge3/bin:/home/lrobidou/miniforge3/condabin:/home/lrobidou/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/lrobidou/.local/bin:/home/lrobidou/.local/bin" VSLANG="1033" "cc" "-Wl,--version-script=/tmp/rustc5FnIWT/list" "-Wl,--no-undefined-version" "-m64" "/tmp/rustc5FnIWT/symbols.o" "-Wl,--as-needed" "-Wl,-Bstatic" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbytecount-68dfcab22851b95b.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbuffer_redux-7b4c530c02a6a1d5.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libmemchr-1df6418796d90927.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd-d4af044339cab2c4.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd_safe-b024a7887f9ab28a.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd_sys-9f924567fce6184b.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma-53c7c9da0626f27c.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma_sys-c180c694e427ca62.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libflate2-fb11999c1a5b09bd.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libminiz_oxide-1a041dcadce55b49.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libadler2-cb22019072e7c439.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libcrc32fast-ef8def920a2f839e.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libcfg_if-181dde3da22bfac8.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2-8dea9fe18d25e2b3.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2_sys-ce23cb44da98297e.rlib" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/liblibc-d1c61429ba3f4f30.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/libstd.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/libkani.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/libkani_core-1876bed1a87ffb66.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-36083ba37d5b6573.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_abort-b4321261d93a754a.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-2c81376a68fbbddd.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-b1fd1a83f089ace4.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-9da743847cda0f38.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-8acd7564d16aebe8.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-3777c8ef8024f93f.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-bad968be82b8ebec.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-a4659070f5b4f9bb.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-6dfdf0a13a372e08.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-d312f2970ac94c6e.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-dba7c3b225206d97.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-662cd2ca1326d643.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-c1872ccfba8ed699.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-f3716bc06e23ec43.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-0a47f69e5c37dd51.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-f97e2f45280f712d.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-8097e946b385dc38.rlib" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c2c57d6302607876.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-B/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/lrobidou/.rustup/toolchains/nightly-2024-12-15-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/lrobidou/.kani/kani-0.57.0/lib" "-L" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/bzip2-sys-b726842da2f7dd95/out/lib" "-L" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/liblzma-sys-7733c1c74cb39892/out" "-L" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/zstd-sys-1e32c7efbbb876af/out" "-L" "/home/lrobidou/.kani/kani-0.57.0/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libneedletail.so" "-Wl,--gc-sections" "-shared" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: rust-lld: error: version script assignment of 'global' to symbol 'bz_internal_error' failed: symbol not defined
          collect2: error: ld returned 1 exit status
          

error: could not compile `needletail` (lib) due to 1 previous error

Caused by:
  process didn't exit successfully: `/home/lrobidou/.kani/kani-0.57.0/bin/kani-compiler --crate-name needletail --edition=2021 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=238 --crate-type cdylib --crate-type rlib --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 --cfg 'feature="bzip2"' --cfg 'feature="compression"' --cfg 'feature="default"' --cfg 'feature="flate2"' --cfg 'feature="liblzma"' --cfg 'feature="xz2"' --cfg 'feature="zstd"' --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("bzip2", "compression", "default", "flate2", "liblzma", "pyo3", "python", "python_test", "xz2", "zstd"))' -C metadata=8124dbd093344b72 --out-dir /home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps --target x86_64-unknown-linux-gnu -C incremental=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/incremental -L dependency=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps -L dependency=/home/lrobidou/Documents/needletail/target/kani/debug/deps --extern buffer_redux=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbuffer_redux-7b4c530c02a6a1d5.rlib --extern bytecount=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbytecount-68dfcab22851b95b.rlib --extern bzip2=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libbzip2-8dea9fe18d25e2b3.rlib --extern flate2=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libflate2-fb11999c1a5b09bd.rlib --extern liblzma=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libliblzma-53c7c9da0626f27c.rlib --extern memchr=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libmemchr-1df6418796d90927.rlib --extern zstd=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libzstd-d4af044339cab2c4.rlib -Cllvm-args=--reachability=harnesses -C overflow-checks=on -Z unstable-options -Z trim-diagnostic-paths=no -Z human_readable_cgu_names -Z always-encode-mir --cfg=kani -Z 'crate-attr=feature(register_tool)' -Z 'crate-attr=register_tool(kanitool)' --sysroot /home/lrobidou/.kani/kani-0.57.0 -L /home/lrobidou/.kani/kani-0.57.0/lib --extern kani --extern 'noprelude:std=/home/lrobidou/.kani/kani-0.57.0/lib/libstd.rlib' -C panic=abort -C symbol-mangling-version=v0 -Z panic_abort_tests=yes -Z mir-enable-passes=-RemoveStorageMarkers '--check-cfg=cfg(kani)' --kani-compiler '-Cllvm-args=--check-version=0.57.0 --log-level=debug --assertion-reach-checks' -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/bzip2-sys-b726842da2f7dd95/out/lib -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/liblzma-sys-7733c1c74cb39892/out -L native=/home/lrobidou/Documents/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/zstd-sys-1e32c7efbbb876af/out` (exit status: 101)
2025-01-10T09:10:11.721169Z DEBUG kani_driver: main_failure error=Failed to execute cargo (exit status: 101). Found 1 compilation errors.
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

I reproduced the issue on a Ubuntu 20.04 machine:
```
$ cargo kani
Kani Rust Verifier 0.57.0 (cargo plugin)
    Updating crates.io index
 Downloading crates ...
  Downloaded bytecount v0.6.8
  Downloaded bzip2 v0.4.4
  Downloaded cc v1.2.9
  Downloaded zstd-safe v7.2.1
  Downloaded zstd v0.13.2
  Downloaded liblzma v0.3.5
  Downloaded buffer-redux v1.0.2
  Downloaded pkg-config v0.3.31
  Downloaded bzip2-sys v0.1.11+1.0.8
  Downloaded zstd-sys v2.0.13+zstd.1.5.6
  Downloaded liblzma-sys v0.3.11
   Compiling libc v0.2.169
   Compiling shlex v1.3.0
   Compiling pkg-config v0.3.31
   Compiling zstd-safe v7.2.1
   Compiling adler2 v2.0.0
   Compiling cfg-if v1.0.0
   Compiling memchr v2.7.4
   Compiling bytecount v0.6.8
   Compiling crc32fast v1.4.2
   Compiling miniz_oxide v0.8.2
   Compiling buffer-redux v1.0.2
   Compiling flate2 v1.0.35
   Compiling jobserver v0.1.32
   Compiling cc v1.2.9
   Compiling zstd-sys v2.0.13+zstd.1.5.6
   Compiling bzip2-sys v0.1.11+1.0.8
   Compiling liblzma-sys v0.3.11
   Compiling bzip2 v0.4.4
   Compiling liblzma v0.3.5
   Compiling zstd v0.13.2
   Compiling needletail v0.6.1 (/home/ubuntu/examples/iss3817/needletail)
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/ubuntu/git/kani/target/kani/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/ubuntu/git/kani/target/kani/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/home/ubuntu/.vscode-server/cli/servers/Stable-fabdb6a30b49f79a7aba0f2ad9df9b399473380f/server/bin/remote-cli:/home/ubuntu/.elan/bin:/home/ubuntu/.opam/4.13.1+options/bin:/home/ubuntu/.local/bin:/home/ubuntu/bin:/home/ubuntu/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ubuntu/.dotnet/tools:/home/ubuntu/git/aws-viewer-for-cbmc:/home/ubuntu/git/kani/scripts:/home/ubuntu/git/kissat/build:/home/ubuntu/git/cadical/build:/home/ubuntu/git/cryptominisat/build:/home/ubuntu/git/smack-deps/corral:/home/ubuntu/sources/dafny:/home/ubuntu/git/aeneas/charon/bin:/home/ubuntu/git/aeneas/bin:/home/ubuntu/git/aws-viewer-for-cbmc:/home/ubuntu/git/kani/scripts:/home/ubuntu/git/kissat/build:/home/ubuntu/git/cadical/build:/home/ubuntu/git/cryptominisat/build:/home/ubuntu/git/smack-deps/corral:/home/ubuntu/sources/dafny:/home/ubuntu/git/aeneas/charon/bin:/home/ubuntu/git/aeneas/bin" VSLANG="1033" "cc" "-Wl,--version-script=/tmp/rustcvCfGyQ/list" "-Wl,--no-undefined-version" "-m64" "/tmp/rustcvCfGyQ/symbols.o" "-Wl,--as-needed" "-Wl,-Bstatic" "/home/ubuntu/examples/iss3817/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/{libbytecount-8998d04d38c201d7.rlib,libbuffer_redux-edd3b54525914a18.rlib,libmemchr-82cc963675f4656e.rlib,libzstd-1e19d6353ad75dcd.rlib,libzstd_safe-807001faa47eda73.rlib,libzstd_sys-dc0f4df344ae264f.rlib,libliblzma-c5330ce4c1db8a7c.rlib,libliblzma_sys-241b7a6ce795140c.rlib,libflate2-dc2ac57700a85b33.rlib,libminiz_oxide-68cc70e3d6cb48ea.rlib,libadler2-983418581eb81d47.rlib,libcrc32fast-4a479b9a8d303de4.rlib,libcfg_if-f4bcd9bf2abe59cc.rlib,libbzip2-e88d649d5e5c8e48.rlib,libbzip2_sys-a41d527c88554778.rlib,liblibc-0fe9aeab35b79bb4.rlib}" "/home/ubuntu/git/kani/target/kani/lib/{libstd.rlib,libkani.rlib,libkani_core-fc1734bcfc9f1b44.rlib}" "/home/ubuntu/git/kani/target/kani/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-af83dd16bb09577c.rlib,libpanic_abort-1f7c5c59003d4dd7.rlib,libobject-cb22f3d4832a6f19.rlib,libmemchr-1488d656a3c6192c.rlib,libaddr2line-65de8a16e89a02fd.rlib,libgimli-645b3ac566a8eb27.rlib,librustc_demangle-22285d2ce0678f45.rlib,libstd_detect-8ca519e0144d4de2.rlib,libhashbrown-6bbf97ebd1b56e57.rlib,librustc_std_workspace_alloc-948e37076470b365.rlib,libminiz_oxide-bd0b8f43df75a0fd.rlib,libadler-37ca75975348312d.rlib,libunwind-0bd7686317c353a4.rlib,libcfg_if-8e47a403c528de00.rlib,liblibc-037284e575351fd5.rlib,liballoc-b832763f58843618.rlib,librustc_std_workspace_core-f6634eb52a751bc9.rlib,libcore-544add90b5a36bd2.rlib,libcompiler_builtins-db444bb6ef9c4620.rlib}" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-B/home/ubuntu/git/kani/target/kani/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B/home/ubuntu/.rustup/toolchains/nightly-2024-12-19-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/ubuntu/git/kani/target/kani/lib" "-L" "/home/ubuntu/examples/iss3817/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/bzip2-sys-f0b4e9060aeb7e3d/out/lib" "-L" "/home/ubuntu/examples/iss3817/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/liblzma-sys-b2202fd0856b8422/out" "-L" "/home/ubuntu/examples/iss3817/needletail/target/kani/x86_64-unknown-linux-gnu/debug/build/zstd-sys-0c14d51c2ddbe0e5/out" "-L" "/home/ubuntu/git/kani/target/kani/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/ubuntu/examples/iss3817/needletail/target/kani/x86_64-unknown-linux-gnu/debug/deps/libneedletail.so" "-Wl,--gc-sections" "-shared" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: version script assignment of 'global' to symbol 'bz_internal_error' failed: symbol not defined
          collect2: error: ld returned 1 exit status
          

error: could not compile `needletail` (lib) due to 1 previous error
error: Failed to execute cargo (exit status: 101). Found 1 compilation errors.
```

--- a/kani-driver/src/call_single_file.rs
+++ b/kani-driver/src/call_single_file.rs
@@ -181,6 +181,8 @@ impl KaniSession {
                 "-Z",
                 "mir-enable-passes=-RemoveStorageMarkers",
                 "--check-cfg=cfg(kani)",
+                // Do not invoke the linker since the compiler will not generate real object files
+                "-Clinker=echo",
             ]
             .map(OsString::from),
         );

--- a/tests/cargo-kani/issue-3817/Cargo.toml
+++ b/tests/cargo-kani/issue-3817/Cargo.toml
@@ -1,0 +1,17 @@
+# Copyright Kani Contributors
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+
+[package]
+name = "issue-3817"
+version = "0.1.0"
+edition = "2024"
+description = "Issue with linking cdylib and another shared library"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[dependencies]
+bzip2 = "0.5.0"
+
+[profile.release]
+lto = true

--- a/tests/cargo-kani/issue-3817/expected
+++ b/tests/cargo-kani/issue-3817/expected
@@ -1,0 +1,9 @@
+Checking harness check_unreachable_extern_fn...
+VERIFICATION:- SUCCESSFUL
+
+Checking harness check_missing_extern_fn...
+Failed Checks: call to foreign "C" function `BZ2_bzCompressInit` is not currently supported by Kani.
+VERIFICATION:- FAILED
+
+Verification failed for - check_missing_extern_fn
+1 successfully verified harnesses, 1 failures, 2 total

--- a/tests/cargo-kani/issue-3817/src/lib.rs
+++ b/tests/cargo-kani/issue-3817/src/lib.rs
@@ -1,0 +1,24 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Check that Kani can compile a crate that depends on bzip, and the analysis will only
+//! fail if a missing symbol is reachable.
+
+use bzip2::Compression;
+use bzip2::read::BzEncoder;
+
+#[kani::proof]
+fn check_missing_extern_fn() {
+    // Call bzip compressor
+    let data: [u8; 10] = kani::any();
+    let compressor = BzEncoder::new(&data[..], Compression::best());
+    assert_eq!(compressor.total_in(), data.len().try_into().unwrap());
+}
+
+#[kani::proof]
+fn check_unreachable_extern_fn() {
+    let positive = kani::any_where(|v: &i8| *v > 0);
+    if positive == 0 {
+        // This should be unreachable so verification should succeed.
+        check_missing_extern_fn();
+    }
+}

